parser/prism/Translator.cc

```diff
@@ -890,6 +891,139 @@ unique_ptr<parser::Node> Translator::translate(pm_node_t *node, bool preserveCon
+                // If the kwargs hash is not present, make a `nil` to put in the place of that argument.
+                // This will be used in the implementation of the intrinsic to tell the difference between keyword
+                // args, keyword args with kw splats, and no keyword args at all.
+                ExpressionPtr kwargsExpr;
+                if (kwargsHash != nullptr) {
+                    ast::Array::ENTRY_store kwargElements;
+                    flattenKwargs(kwargsHash, kwargElements);
+                    ast::desugar::DuplicateHashKeyCheck::checkSendArgs(ctx, 0, kwargElements);
+
+                    // Add the kwargs Hash back into parse tree, so that it's correct, too.
+                    // This doesn't effect the desugared expression.
+                    args.emplace_back(move(kwargsHash));
+
+                    kwargsExpr = MK::Array(loc, move(kwargElements));
+                } else {
+                    kwargsExpr = MK::Nil(loc);
+                }
+
+                if (prismBlock && PM_NODE_TYPE_P(prismBlock, PM_BLOCK_ARGUMENT_NODE)) {
+                    // Add the parser node back into the wq tree, to pass the parser tests.
+                    args.emplace_back(move(blockPassNode));
+                }
+
+                auto numPosArgs = 4;
+                ast::Send::ARGS_store magicSendArgs;
+                magicSendArgs.reserve(numPosArgs); // TODO: reserve room for a block pass arg
+                magicSendArgs.emplace_back(move(receiverExpr));
+                magicSendArgs.emplace_back(move(methodName));
+                magicSendArgs.emplace_back(move(argsArrayExpr));
+                magicSendArgs.emplace_back(move(kwargsExpr));
+
+                if (prismBlock != nullptr && PM_NODE_TYPE_P(prismBlock, PM_BLOCK_ARGUMENT_NODE) &&
+                    !blockPassArgIsSymbol) {
+                    // Special handling for non-Symbol block pass args, like `a.map(&block)`
+                    // Symbol procs like `a.map(:to_s)` are rewritten into literal block arguments,
+                    // and handled separately below.
+
+                    // Desugar a call with a splat, and any other expression as a block pass argument.
+                    // E.g. `foo(*splat, &block)`
+
+                    magicSendArgs.emplace_back(move(blockPassArg));
+                    numPosArgs++;
+
+                    auto sendExpr = MK::Send(loc, MK::Magic(loc), core::Names::callWithSplatAndBlockPass(), messageLoc,
+                                             numPosArgs, move(magicSendArgs), flags);
+                    return make_node_with_expr<parser::Send>(move(sendExpr), loc, move(receiver), name, messageLoc,
+                                                             move(args));
+                }
+
+                if (prismBlock != nullptr) {
+                    if (PM_NODE_TYPE_P(prismBlock, PM_BLOCK_NODE) || blockPassArgIsSymbol) {
```

Comment:
This block is analogous to these two cases:

https://github.com/sorbet/sorbet/blob/13459ea30ab9e3ac2ead5ca4411549ddd9940131/ast/desugar/Desugar.cc#L855-L871